### PR TITLE
[SelectionDAG] Propagate Disjoint flag.

### PR DESCRIPTION
**Pull request by fengfeng09**

### CODE REVIEW

**Review by RKSimon — COMMENTED** [review]
Tests?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
Please add some tests for these changes.

**Comment by fengfeng09** [comment]
Thanks all, updated.

File: llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll [hunk 1]
```diff
@@ -0,0 +1,22 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+; RUN: llc -mtriple=x86_64 %s -start-before=x86-isel -o - -stop-after=x86-isel | FileCheck %s
+
+define void @add_shl_or_disjoint(i32 %x, ptr addrspace(1) %o) {
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll:4]: I don't think this test should be using MIR. If there's an observable codegen change, this should just check the final ISA. A MIR test would be the right choice if it was only showing the flag survived through selection, but I don't see it here

Comment (fengfeng09) [reply llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll:4]: This commit actually aimed at propagating the disjoint flag in DAGCombine. I choose MIR test just want to check the result of the seletion, no any other concern or consideration. Maybe ll test is more suitable as your suggestion.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -9530,6 +9530,8 @@ static SDValue combineShiftOfShiftedLogic(SDNode *Shift, SelectionDAG &DAG) {
   SDValue ShiftSumC = DAG.getConstant(*C0Val + C1Val, DL, ShiftAmtVT);
   SDValue NewShift1 = DAG.getNode(ShiftOpcode, DL, VT, X, ShiftSumC);
   SDValue NewShift2 = DAG.getNode(ShiftOpcode, DL, VT, Y, C1);
+  // Propogate flags.
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:9533]: Drop comment

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 3]
```diff
@@ -544,6 +544,8 @@ bool TargetLowering::ShrinkDemandedConstant(SDValue Op,
     if (!C.isSubsetOf(DemandedBits)) {
       EVT VT = Op.getValueType();
       SDValue NewC = TLO.DAG.getConstant(DemandedBits & C, DL, VT);
+      // Propogate flags.
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:547]: Drop comment

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+; CHECK-LABEL: add_shl_or_disjoint:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    # kill: def $edi killed $edi def $rdi
+; CHECK-NEXT:    leal 165036822(,%rdi,4), %eax
+; CHECK-NEXT:    movl %eax, (%rsi)
+; CHECK-NEXT:    retq
+  %or = or disjoint i32 %x, 4027584529
+  %shl = shl i32 %or, 2
+  %add = add i32 %shl, 1234567890
+  store i32 %add, ptr addrspace(1) %o
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll:14]: Don't use the addrspace(1). Can you just return the value instead?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 2 above)
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:9533]: https://alive2.llvm.org/ce/z/ttZfv8

File: llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll (hunk 4 above)
Comment (fengfeng09) [reply llvm/test/CodeGen/X86/propogate-disjoint-in-shl-or.ll:14]: ok

**Comment by jayfoad** [comment]
Typo in subject "propagate"

**Review by dtcxzyw — APPROVED** [review]
LGTM. Nice catch!

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 5]
```diff
@@ -9530,6 +9530,7 @@ static SDValue combineShiftOfShiftedLogic(SDNode *Shift, SelectionDAG &DAG) {
   SDValue ShiftSumC = DAG.getConstant(*C0Val + C1Val, DL, ShiftAmtVT);
   SDValue NewShift1 = DAG.getNode(ShiftOpcode, DL, VT, X, ShiftSumC);
   SDValue NewShift2 = DAG.getNode(ShiftOpcode, DL, VT, Y, C1);
+  SelectionDAG::FlagInserter FlagsInserter(DAG, LogicOp->getFlags());
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:9533]: Why do we need to use FlagInserter? Can't we pass the flags to the last argument of getNode?

**Comment by fengfeng09** [comment]
I have no write access to the repo，could someone help me merge this PR？

**Review by arsenm — APPROVED** [review]

**Pull request merged**